ToDo.Droid/AndroidViewService.cs

```diff
@@ -0,0 +1,66 @@
+﻿using ToDo.Core.Services;
+using ToDo.UI;
+using Xamarin.Essentials;
+using Xamarin.Forms;
+
+namespace ToDo.Droid
+{
+    public class AndroidViewService : INativeViewService
+    {
+        public int[] GetCoordinates(VisualElement visualElement)
+        {
+            // Get the position of the StackLayout
+            var stackLayout = (StackLayout)visualElement.Parent.Parent.Parent;
+
+            var view = Xamarin.Forms.Platform.Android.Platform.GetRenderer(stackLayout).View;
+
+            var coords = new int[4];
+            view.GetLocationOnScreen(coords);
+
+            var density = view.Context.Resources.DisplayMetrics.Density;
+
+            var stackLayoutY = (int)(coords[1] / density);
+
+            // Get the position of the selected grid
+            view = Xamarin.Forms.Platform.Android.Platform.GetRenderer(visualElement).View;
+
+            view.GetLocationOnScreen(coords);
+
+            for (int i = 0; i < 4; i++)
+            {
+                coords[i] = (int)(coords[i] / density);
+            }
+
+            coords[1] = coords[1] - stackLayoutY;
+
+            return coords;
+        }
+
+        public int ValidateYPosition(int y, VisualElement grid)
+        {
+            var height = (int)(DeviceDisplay.MainDisplayInfo.Height / DeviceDisplay.MainDisplayInfo.Density);
+
+            /*
+            // 1. Parent : ViewCell
+            // 2. Parent : ListView
+            // 3. Parent : StackLayout
+            var stackLayout = (StackLayout)grid.Parent.Parent.Parent;
+
+            // 20 : margin (10 above + 10 below)
+            var buttonHeight = (int)stackLayout.Children[2].Bounds.Height + 20;
+            */
+
+            // Edit/Delete row : 60
+            // Done/Pending row : 60
+            // Close button row : 140
+            // +5 to add a little space between the Close icon and the bottom of the screen
+            // Sum : 265
+            if (y > (height - ((int)grid.Bounds.Height + 365)))
```

Comment:
Not really. 😞 I'm totally not proud of this part of the solution. 😐 I spent a lot of time trying to figure out the right height. The grids have a fix height, I sumed them, that's what you can see in the comments and in my very precise sketch 😄 (where the red line is the border of the screen). But it didn't work (the value was too small), so I thought maybe I should add the height of the "New" button as well (you can see my tries in the commented code), but it didn't work either, this value was still too small so the close button stuck out of the screen.
At this point i gave up and guessed some bigger value. 360 was the exactly right value, then i added 5 more to let a little margin between the bottom of the button and the bottom of the screen.
That 360 doesn't add up to me. Then i tried the app in iOS and the grid was not in the right position, there was a big space between the close icon and the bottom of the screen. So i guessed a smaller value (which is 295 + 5 on iOS). I know, this "I don't understand why but it works" approach is not good but I didn't want to spend days with this. 😐 I still don't understand what is the problem with my calculation, it seems logical to me.

![image](https://user-images.githubusercontent.com/44465559/98538025-734a1c00-228a-11eb-9f70-8a03eaefde42.png)

Reply:
Nice sketch! :smile: I think your calculations are right, but did you also try to debug that and see what height the outer container of the while thing has? I suppose there is a Grid or StackLayout all around this?

I support your judgement that the effort is not really worth it, since your solution already works. It would be nice to reuse that service out of the box, even with a different layout, but that's not a blocker. However, you can write a comment to explain why this value is there, so that your colleagues or even you in 6 months can understand what is going on. We use this approach a lot when code is not self-explaining.

Reply:
They are in a Grid and I forgot to think about that..🤦 I'll check it again and debug it later.